python/pyarrow/tests/test_udf.py

```diff
@@ -0,0 +1,484 @@
+        pc.register_scalar_function(add_const, "test_no_desc",
+                                    func_doc, in_types,
+                                    out_type)
+
+    # doc with empty dictionary
+    func_doc = {}
+    expected_expr = "Function doc must contain a summary"
+    with pytest.raises(ValueError, match=expected_expr):
+        pc.register_scalar_function(add_const,
+                                    "test_empty_dictionary",
+                                    func_doc, in_types,
+                                    out_type)
+
+
+def test_nullary_functions(nullary_check_func_fixture, mock_udf_context):
+    check_scalar_function(nullary_check_func_fixture, [],
+                          mock_udf_context, False)
+
+
+def test_output_datatype(output_check_func_fixture):
+    function, func_name = output_check_func_fixture
+    func = pc.get_function(func_name)
+
+    assert func.name == func_name
+
+    expected_expr = "Expected output type, int64," \
+        + " but function returned type int32"
+
+    with pytest.raises(pa.lib.ArrowTypeError, match=expected_expr):
+        pc.call_function(func_name, [pa.array([20, 30])])
+
+
+def test_output_value():
+    def get_output(ctx, array):
+        return pc.call_function("add", [array, 1])
+    func_name = "test_output_value"
+    in_types = {"array": pa.int64()}
+    out_type = {}
+    doc = {
+        "summary": "test output value",
+        "description": "test output"
+    }
+
+    expected_expr = "DataType expected, got <class 'dict'>"
+
+    with pytest.raises(TypeError, match=expected_expr):
+        pc.register_scalar_function(output_check_func_fixture, func_name, doc,
+                                    in_types, out_type)
+
+
+def test_output_type(output_type_func_fixture):
+    _, func_name = output_type_func_fixture
+
+    func = pc.get_function(func_name)
+
+    assert func.name == func_name
+
+    expected_expr = "Unexpected output type: int"
+
+    with pytest.raises(pa.lib.ArrowTypeError, match=expected_expr):
+        pc.call_function(func_name, [pa.array([20, 30])])
+
+
+def test_input_type():
+    func_name = "test_input_type"
+    in_types = {"array": None}
+    out_type = pa.int64()
+    doc = {
+        "summary": "test invalid input type",
+        "description": "invalid input function"
+    }
+    expected_expr = "in_types must be of type DataType"
+
+    with pytest.raises(TypeError, match=expected_expr):
+        pc.register_scalar_function(const_return, func_name, doc,
+                                    in_types, out_type)
+
+
+def test_udf_context(random_with_udf_ctx_func_fixture):
+    proxy_pool = pa.proxy_memory_pool(pa.default_memory_pool())
+    _, func_name = random_with_udf_ctx_func_fixture
+
+    res = pc.call_function(func_name,
+                           [pa.scalar(10), pa.scalar(20)],
+                           memory_pool=proxy_pool)
+    assert res[0].as_py() == 30
+    assert proxy_pool.bytes_allocated() == 64
```

Comment:
This is brittle, just assert that it is nonzero.

Reply:
Ah, sure, let's do that.